### DISCUSSION TITLE
custom storage/backends

### DISCUSSION THREAD

**Discussion by c-antin** [GitoxideLabs/gitoxide#1281] [Q&A]
### Summary 💡

Any plans on supporting custom backends?
See https://github.com/go-git/go-git/tree/v5.11.0/storage
Or https://github.com/libgit2/libgit2-backends

### Motivation 🔦

Use case: in-memory and/or database storage for objectDB, refDB, etc. for faster access or reduced storage footprint.

**Comment by Byron** [comment]
Currently there are traits available for reading from and writing to the object database, but the main implementation can't yet be replaced.
It should be possible to one day put it behind a trait and thus allow alternative implementations, but it's not on the roadmap just yet, so I'd think stabilization happens before it.

Questions related to this are WASM support, as currently `gitoxide` can't compile to it for the most part, but will have better chances with swappable object storage backends.

**Comment by yumin-chen** [comment]
isomorphic-gitoxide

**Reply by Byron** [reply]
While this probably is a hint at https://www.npmjs.com/package/isomorphic-git, I still don't get it.

Besides that, I may share the news that there now is [a way](https://docs.rs/gix/latest/gix/struct.Repository.html#method.with_object_memory) to keep newly created objects in memory.

Similar facilities will ideally be possible for references as well once `gix-ref` has a way to represent everything behind a trait, allowing a layering of database, one in memory, the other one being a normal backend.
